resilience4j-spring/src/main/java/io/github/resilience4j/decorators/configure/DecoratorsAspect.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2019 Mahmoud Romeh
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.github.resilience4j.decorators.configure;
+
+import io.github.resilience4j.bulkhead.annotation.Bulkhead;
+import io.github.resilience4j.bulkhead.configure.BulkheadAspectHelper;
+import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker;
+import io.github.resilience4j.circuitbreaker.configure.CircuitBreakerAspectHelper;
+import io.github.resilience4j.retry.configure.*;
+import java.util.List;
+
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.aspectj.lang.annotation.Pointcut;
+import org.springframework.core.Ordered;
+
+import io.github.resilience4j.core.lang.Nullable;
+import io.github.resilience4j.decorators.annotation.Decorator;
+import io.github.resilience4j.ratelimiter.annotation.RateLimiter;
+import io.github.resilience4j.ratelimiter.configure.RateLimiterAspectHelper;
+import io.github.resilience4j.retry.annotation.Retry;
+import io.github.resilience4j.utils.ProceedingJoinPointHelper;
+import java.util.Arrays;
+import java.util.Collections;
+
+/**
+ * This Spring AOP aspect intercepts all methods which are annotated with a {@link Decorator}
+ * annotation. The aspect will handle methods that return a RxJava2 reactive type, Spring Reactor
+ * reactive type, CompletionStage type, or value type.
+ *
+ * The RetryRegistry is used to retrieve an instance of a Retry for a specific name.
+ *
+ * Given a method like this:
+ * <pre><code>
+ *     {@literal @}Decorator(retry = @Retry(name = "readTimeouts"))
```

Comment:
How can we guarantee that the order of the `@Decorator` annotations is applied correctly.
I think it's not ensured that the Aspects executes them in that order, or?

Reply:
there is only one aspect that handles the `Decorator` annotations: `DecoratorsAspect`
it extracts an order list of `Decorator` annotations on a given method (or class) and decorates the proceeding call with respect to the order in which those 

see `DecoratorsAspect` implementations

Reply:
Yes, but as far as I remember: When a method has same annotation multiple times, Spring AOP does not guarantee that the Aspect processes them exactly in the order they are specified.

Reply:
I might be wrong. We have to test it.
That's why I always thought we have to use one annotation to combine them

```
@Decorators(value = [
    @Retry(name = "tickerAllPrices"), 
    @CircuitBreaker(name = "tickerAllPrices"),
    @RateLimiter(name = "tickerAllPrices")    
])
```

Reply:
to my understanding Spring AOP checks if this criteria is true:

```
    @Pointcut(
        value = "@within(decorator) || @annotation(decorator)",
        argNames = "decorator")
    public void matchAnnotatedClassOrMethod(Decorator decorator) {
    }

    @Around(
        value = "matchAnnotatedClassOrMethod(foundDecoratorAnnotation)",
        argNames = "proceedingJoinPoint, foundDecoratorAnnotation")
```

and if it is true then it will call `decorateAroundAdvice` once for the given method call even if it has many `@Decorator` annotations.

I'll test this today in the debugger and add a junit test later.

As for you proposal

```
@Decorators(value = [
    @Retry(name = "tickerAllPrices"), 
    @CircuitBreaker(name = "tickerAllPrices"),
    @RateLimiter(name = "tickerAllPrices")    
])
```

I don't think this is doable because Java does not support annotation inheritance and `@Retry`, `@CircuitBreaker` and `@RateLimiter` would have to have a common super class to make this work.

Reply:
Ok, I missed out `ProceedingJoinPointHelper`.
Could you please check if the Apsect is only invoked once or multiple times?

Reply:
guess what, when I've added a second `@Decorator` to a method it got called zero times :o)

if I add a second `@Decorator` the AOP does not see it as a `Decorator` but a `Decorators` annotation holding those two `Decorator` instances.

Pushed a commit that fixes this and now this:

```
@Component
public class RetryDummyService implements TestDummyService {

    @Override
    @Decorator(rateLimiter = @RateLimiter(name = BACKEND))
    @Decorator(retry = @Retry(name = BACKEND, fallbackMethod = "recovery"))
    public String sync() {
        return syncError();
    }
```

results in `decorateAroundAdvice` getting called once.

I'll refactor `DecoratorsAspect` to a single pointcuts. However in current form it proves that my annotation scheme is implementable.

Reply:
Very nice. Thanks